File: src/physics.ts
import { vec3 } from "gl-matrix";
import { Camera } from "./camera";
import { World } from "./world/world";
import Time from "./time-manager";
import { BlockType } from "./world/block";

const GRAVITY = -28.0; // units/s²
const JUMP_SPEED = 9.0; // units/s upward on jump
const WALK_SPEED = 5.0; // units/s horizontal
const TERMINAL_VEL = -50.0; // minimum Y velocity
const EYE_HEIGHT = 1.6; // camera Y above feet
const HALF_W = 0.3; // player AABB half-width (X and Z)
const PLAYER_HEIGHT = 1.8; // player AABB full height (Y)

/**
 * Axis-aligned bounding-box physics for the player.
 *
 * Each frame, gravity is applied to Y velocity and then the player
 * is swept along each axis independently. If a sweep overlaps a solid
 * block the movement is undone and that velocity component is zeroed.
 * This prevents tunneling while keeping the implementation simple.
 */
export class Physics {
  /** World-space velocity in units/s. */
  velocity: vec3 = vec3.create();

  /** True when the player is standing on a solid surface. */
  onGround = false;

  /** Feet position – camera is at feet + EYE_HEIGHT. */
  private feet: vec3;

  constructor(
    private camera: Camera,
    private world: World,
  ) {
    // Derive feet from the camera's starting position.
    this.feet = vec3.fromValues(
      camera.position[0],
      camera.position[1] - EYE_HEIGHT,
      camera.position[2],
    );
  }

  /** Attempt a jump; only succeeds when standing on the ground. */
  jump(): void {
    if (this.onGround) {
      this.velocity[1] = JUMP_SPEED;
      this.onGround = false;
    }
  }

  /**
   * Advance physics by one frame.
   * @param wishX  Desired horizontal X velocity direction (−1 … 1, pre-normalised).
   * @param wishZ  Desired horizontal Z velocity direction (−1 … 1, pre-normalised).
   */
  update(wishX: number, wishZ: number): void {
    // Cap delta so that a long pause (tab hidden, breakpoint, etc.)
    // can't teleport the player through a block.
    const dt = Math.min(Time.deltaTime, 0.05);

    // ── Gravity (always applied so the player is probed down every frame) ──
    this.velocity[1] += GRAVITY * dt;
    if (this.velocity[1] < TERMINAL_VEL)
      this.velocity[1] = TERMINAL_VEL;

    // ── Horizontal wish velocity (instant, no friction sim for now) ────────
    this.velocity[0] = wishX * WALK_SPEED;
    this.velocity[2] = wishZ * WALK_SPEED;

    this.onGround = false;

    // ── Sweep X ─────────────────────────────────────────────────────────────
    this.feet[0] += this.velocity[0] * dt;
    if (this.collidesWorld()) {
      this.feet[0] -= this.velocity[0] * dt;
      this.velocity[0] = 0;
    }

    // ── Sweep Y ─────────────────────────────────────────────────────────────
    const velY = this.velocity[1];
    this.feet[1] += velY * dt;
    if (this.collidesWorld()) {
      this.feet[1] -= velY * dt;
      if (velY < 0) this.onGround = true; // hit something below → grounded
      this.velocity[1] = 0;
    }

    // ── Sweep Z ─────────────────────────────────────────────────────────────
    this.feet[2] += this.velocity[2] * dt;
    if (this.collidesWorld()) {
      this.feet[2] -= this.velocity[2] * dt;
      this.velocity[2] = 0;
    }

    // ── Sync camera to feet position ─────────────────────────────────────
    this.camera.position[0] = this.feet[0];
    this.camera.position[1] = this.feet[1] + EYE_HEIGHT;
    this.camera.position[2] = this.feet[2];
  }

  /**
   * Returns true if the player AABB (at the current feet position) overlaps
   * any solid block in the world.
   */
  private collidesWorld(): boolean {
    const minX = this.feet[0] - HALF_W;
    const maxX = this.feet[0] + HALF_W;
    const minY = this.feet[1];
    const maxY = this.feet[1] + PLAYER_HEIGHT;
    const minZ = this.feet[2] - HALF_W;
    const maxZ = this.feet[2] + HALF_W;

    // Integer block range that the AABB can touch.
    const bx0 = Math.floor(minX);
    const bx1 = Math.floor(maxX);
    const by0 = Math.floor(minY);
    const by1 = Math.floor(maxY);
    const bz0 = Math.floor(minZ);
    const bz1 = Math.floor(maxZ);

    for (let bx = bx0; bx <= bx1; bx++) {
      for (let by = by0; by <= by1; by++) {
        for (let bz = bz0; bz <= bz1; bz++) {
          if (
            this.world.getBlock(bx, by, bz) !==
            BlockType.Air
          ) {
            return true;
          }
        }
      }
    }
    return false;
  }

  getState(): {
    velocity: [number, number, number];
    onGround: boolean;
  } {
    return {
      velocity: [
        this.velocity[0],
        this.velocity[1],
        this.velocity[2],
      ],
      onGround: this.onGround,
    };
  }

  setState(state: {
    velocity: [number, number, number];
    onGround: boolean;
  }): void {
    this.velocity[0] = state.velocity[0];
    this.velocity[1] = state.velocity[1];
    this.velocity[2] = state.velocity[2];
    this.onGround = state.onGround;
    this.feet[0] = this.camera.position[0];
    this.feet[1] = this.camera.position[1] - EYE_HEIGHT;
    this.feet[2] = this.camera.position[2];
  }
}

File: game/src/engine/physics.rs
use glam::Vec3;

use crate::engine::camera::Camera;
use crate::world::block::BlockType;
use crate::world::World;

const GRAVITY: f32 = -28.0; // units/s²
const JUMP_SPEED: f32 = 9.0; // units/s upward
const WALK_SPEED: f32 = 5.0; // units/s horizontal
const TERMINAL_VEL: f32 = -50.0; // min Y velocity
const EYE_HEIGHT: f32 = 1.6; // camera Y above feet
const HALF_W: f32 = 0.3; // player AABB half-width
const PLAYER_HEIGHT: f32 = 1.8; // player AABB height

pub struct Physics {
    pub velocity: Vec3,
    pub on_ground: bool,
    /// Feet position; camera = feet + EYE_HEIGHT.
    feet: Vec3,
}

impl Physics {
    pub fn new(camera: &Camera) -> Self {
        let feet = Vec3::new(
            camera.position.x,
            camera.position.y - EYE_HEIGHT,
            camera.position.z,
        );
        Self {
            velocity: Vec3::ZERO,
            on_ground: false,
            feet,
        }
    }

    /// Attempt a jump; only when standing on ground.
    pub fn jump(&mut self) {
        if self.on_ground {
            self.velocity.y = JUMP_SPEED;
            self.on_ground = false;
        }
    }

    /// Sync the physics feet position to match the
    /// camera (call after directly moving the camera,
    /// e.g. while flying).
    pub fn sync_to_camera(&mut self, camera: &Camera) {
        self.feet = Vec3::new(
            camera.position.x,
            camera.position.y - EYE_HEIGHT,
            camera.position.z,
        );
    }

    /// Advance physics one frame.
    ///
    /// `wish_x` / `wish_z` — desired horizontal
    /// movement direction (pre-computed from camera
    /// yaw + key input, normalised).
    pub fn update(
        &mut self,
        wish_x: f32,
        wish_z: f32,
        dt: f32,
        world: &World,
        camera: &mut Camera,
    ) {
        // Cap dt to avoid tunnelling through blocks
        // on long pauses / breakpoints.
        let dt = dt.min(0.05);

        // Gravity
        self.velocity.y += GRAVITY * dt;
        if self.velocity.y < TERMINAL_VEL {
            self.velocity.y = TERMINAL_VEL;
        }

        // Horizontal wish velocity
        self.velocity.x = wish_x * WALK_SPEED;
        self.velocity.z = wish_z * WALK_SPEED;

        self.on_ground = false;

        // Sweep X
        self.feet.x += self.velocity.x * dt;
        if self.collides_world(world) {
            self.feet.x -= self.velocity.x * dt;
            self.velocity.x = 0.0;
        }

        // Sweep Y
        let vel_y = self.velocity.y;
        self.feet.y += vel_y * dt;
        if self.collides_world(world) {
            self.feet.y -= vel_y * dt;
            if vel_y < 0.0 {
                self.on_ground = true;
            }
            self.velocity.y = 0.0;
        }

        // Sweep Z
        self.feet.z += self.velocity.z * dt;
        if self.collides_world(world) {
            self.feet.z -= self.velocity.z * dt;
            self.velocity.z = 0.0;
        }

        // Sync camera
        camera.position.x = self.feet.x;
        camera.position.y = self.feet.y + EYE_HEIGHT;
        camera.position.z = self.feet.z;
    }

    fn collides_world(&self, world: &World) -> bool {
        let min_x = self.feet.x - HALF_W;
        let max_x = self.feet.x + HALF_W;
        let min_y = self.feet.y;
        let max_y = self.feet.y + PLAYER_HEIGHT;
        let min_z = self.feet.z - HALF_W;
        let max_z = self.feet.z + HALF_W;

        let bx0 = min_x.floor() as i32;
        let bx1 = max_x.floor() as i32;
        let by0 = min_y.floor() as i32;
        let by1 = max_y.floor() as i32;
        let bz0 = min_z.floor() as i32;
        let bz1 = max_z.floor() as i32;

        for bx in bx0..=bx1 {
            for by in by0..=by1 {
                for bz in bz0..=bz1 {
                    if world.get_block(bx, by, bz) != BlockType::Air {
                        return true;
                    }
                }
            }
        }
        false
    }

    pub fn get_state(&self) -> PhysicsState {
        PhysicsState {
            velocity: self.velocity.to_array(),
            on_ground: self.on_ground,
        }
    }

    pub fn set_state(&mut self, state: &PhysicsState, camera: &Camera) {
        self.velocity = Vec3::from(state.velocity);
        self.on_ground = state.on_ground;
        self.feet = Vec3::new(
            camera.position.x,
            camera.position.y - EYE_HEIGHT,
            camera.position.z,
        );
    }
}

pub struct PhysicsState {
    pub velocity: [f32; 3],
    pub on_ground: bool,
}
